lib/ash_json_api/serializer.ex

```diff
@@ -206,73 +220,164 @@ defmodule AshJsonApi.Serializer do
+        Map.put(query, "page", %{
+          offset: offset
+        })
+
+      true ->
+        Map.put(query, "page", %{
+          limit: limit,
+          offset: offset
+        })
+    end
   end
 
-  defp put_page_params(query, %{limit: limit, offset: offset}) do
-    Map.put(query, "page", %{
-      limit: limit,
-      offset: offset
-    })
-  end
+  defp put_page_params(query, %Ash.Page.Keyset{} = paginator) do
+    %{after: after_cursor, before: before_cursor, limit: limit} = paginator
 
-  defp put_page_params(query, %{after: after_keyset, limit: nil}) when is_binary(after_keyset) do
-    Map.put(query, "page", %{after: after_keyset})
-  end
+    cond do
+      is_nil(limit) ->
+        Map.put(query, "page", %{after: after_cursor})
 
-  defp put_page_params(query, %{after: after_keyset, limit: limit})
-       when is_binary(after_keyset) do
-    Map.put(query, "page", %{after: after_keyset, limit: limit})
-  end
+      is_binary(after_cursor) ->
+        Map.put(query, "page", %{after: after_cursor, limit: limit})
+
+      is_nil(limit) and is_binary(before_cursor) ->
+        Map.put(query, "page", %{before: before_cursor})
+
+      is_binary(before_cursor) ->
+        Map.put(query, "page", %{before: before_cursor, limit: limit})
 
-  defp put_page_params(query, %{before: before_keyset, limit: nil})
-       when is_binary(before_keyset) do
-    Map.put(query, "page", %{before: before_keyset})
+      true ->
+        Map.put(query, "page", %{limit: limit})
+    end
   end
 
-  defp put_page_params(query, %{before: before_keyset, limit: limit})
-       when is_binary(before_keyset) do
-    Map.put(query, "page", %{before: before_keyset, limit: limit})
+  defp put_page_params(query, _), do: query
+
+  # Offset pagination
+
+  defp add_next_link(links, uri, query, %Ash.Page.Offset{} = paginator) do
+    %{results: results, count: count, offset: offset, limit: limit} = paginator
+
+    cond do
+      not is_nil(count) and not is_nil(offset) and offset + limit >= count ->
+        Map.put(links, :next, nil)
+
+      not is_nil(count) and is_nil(offset) and limit >= count ->
+        Map.put(links, :next, nil)
+
+      Enum.count(results) < limit ->
+        Map.put(links, :next, nil)
+
+      true ->
+        new_query =
+          query
+          |> put_page_params(next_page(paginator))
+          |> Conn.Query.encode()
+
+        link =
+          uri
+          |> put_query(new_query)
+          |> URI.to_string()
+          |> encode_link()
+
+        Map.put(links, :next, link)
+    end
   end
 
-  defp put_page_params(query, %{limit: limit}) when is_integer(limit),
-    do: Map.put(query, "page", %{limit: limit})
+  ## Cursor pagination
 
-  defp put_page_params(query, _), do: query
+  defp add_next_link(links, uri, query, %Ash.Page.Keyset{} = paginator) do
+    case paginator do
+      # This is a query for the first page, no cursors are set
+      %{results: results, after: nil, before: nil, more?: true} ->
```

Comment:
we need a case for `after: nil, before: nil, more?: false` I think

Reply:
added